staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -972,31 +1101,41 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer, budget *ti
 	budget.returnUnused(timeout - time.Since(startTime))
 }
 
+func (c *cacheWatcher) nextBookmarkTime(now time.Time) time.Time {
+	// For now we return 2s before deadline (and maybe +infinity is now already passed this time)
+	// but it gives us extensibility for the future.
+	return c.deadline.Add(-2 * time.Second)
+}
+
 // NOTE: sendWatchCacheEvent is assumed to not modify <event> !!!
 func (c *cacheWatcher) sendWatchCacheEvent(event *watchCacheEvent) {
-	curObjPasses := event.Type != watch.Deleted && c.filter(event.Key, event.ObjLabels, event.ObjFields)
-	oldObjPasses := false
-	if event.PrevObject != nil {
-		oldObjPasses = c.filter(event.Key, event.PrevObjLabels, event.PrevObjFields)
-	}
-	if !curObjPasses && !oldObjPasses {
-		// Watcher is not interested in that object.
-		return
-	}
-
 	var watchEvent watch.Event
-	switch {
-	case curObjPasses && !oldObjPasses:
-		watchEvent = watch.Event{Type: watch.Added, Object: event.Object.DeepCopyObject()}
-	case curObjPasses && oldObjPasses:
-		watchEvent = watch.Event{Type: watch.Modified, Object: event.Object.DeepCopyObject()}
-	case !curObjPasses && oldObjPasses:
-		// return a delete event with the previous object content, but with the event's resource version
-		oldObj := event.PrevObject.DeepCopyObject()
-		if err := c.versioner.UpdateObject(oldObj, event.ResourceVersion); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failure to version api object (%d) %#v: %v", event.ResourceVersion, oldObj, err))
+	if event.Type == watch.Bookmark {
+		watchEvent = watch.Event{Type: watch.Bookmark, Object: event.Object.DeepCopyObject()}
+	} else {
+		curObjPasses := event.Type != watch.Deleted && c.filter(event.Key, event.ObjLabels, event.ObjFields)
+		oldObjPasses := false
+		if event.PrevObject != nil {
+			oldObjPasses = c.filter(event.Key, event.PrevObjLabels, event.PrevObjFields)
+		}
+		if !curObjPasses && !oldObjPasses {
```

Comment:
Make this the 1st case in the switch?

Reply:
This is just indentation on the already existing code, which I'm trying to minimize changes on as much as possible.
But probably this one actually makes sense.

Reply:
add a helper function here.